sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/CounterCell.java

```diff
@@ -0,0 +1,25 @@
+package org.apache.beam.sdk.metrics;
+
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
+
+/**
+ * Tracks the current value (and delta) for a Counter metric for a specific context and bundle.
+ */
+@Experimental(Kind.METRICS)
+class CounterCell extends MetricCell<Long> {
+
+  private final AtomicLong value = new AtomicLong();
+
+  /** Increment the counter by the given amount. */
+  public void add(long n) {
+    markDirty();
```

Comment:
I don't know that we need to go that far. I think the right behavior is to call markDirty() after the add(), as you initially hinted at. The race then becomes we may have reported a value that included the added data and then mark dirty and then report it again with no change. But, the dirtiness is just an optimization intended to allow us to report less often than we need, so this isn't a problem.